### PR TITLE
New error discipline

### PR DESCRIPTION
**Pull request by kurtbrose**

### CODE REVIEW

**Comment by kurtbrose** [comment]
alright, got past a major hurdle that apparently came down to sys.exc_info() and traceback.format_exc() seeing different stacks running under pytest (somehow)

punchline: here is what a glom "stack trace" looks like now

```python
E           glom.core.glom.<locals>.GlomWrapError: -> {'results': [{'id': 'id', 'name': 'name', 'add_date': ('add_date', ...     
E           OrderedDict([(1, Contact(id=1, name='Kurt', pref_name='', emails=[Emai...                                            
E           -> [{'id': 'id', 'name': 'name', 'add_date': ('add_date', <class 'str'...                                            
E           -> {'id': 'id', 'name': 'name', 'add_date': ('add_date', <class 'str'>...                                            
E           1                                                                                                                    
E           -> 'id'                                                                                                              
E             File "C:\Users\Kurt\workspace\glom\glom\core.py", line 399, in glomit                                              
E               return _t_eval(target, self.path_t, scope)                                                                       
E             File "C:\Users\Kurt\workspace\glom\glom\core.py", line 1232, in _t_eval                                            
E               raise PathAccessError(e, Path(_t), i // 2) from None                                                             
E           glom.core.PathAccessError: could not access 'id', part 0 of Path('id'), got error: AttributeError("'int' object has n
o attribute 'id'")                                                                                                               
```

**Comment by kurtbrose** [comment]
oops I think I was too aggressive in ripping out the guts of PathAccessError -- it still is very necessary because _t_eval isn't looping back through glom() on each path element

(theoretically we could do so -- basically it would make Path() equivalent to a tuple of the segments more-or-less; rather leave that for the future though, this is a gigantic enough change as it is)

**Comment by kurtbrose** [comment]
dunno why the diff isn't updating, anyway getting really close --

there's some kind of circularity happening with _finalize() and format

```python
glom.core.PathAccessError:
-> 'a.b'
{}
  File "C:\Users\Kurt\workspace\glom\glom\core.py", line 423, in glomit
    return _t_eval(target, self.path_t, scope)
  File "C:\Users\Kurt\workspace\glom\glom\core.py", line 1256, in _t_eval
    raise PathAccessError(e, Path(_t), i // 2) from None
glom.core.PathAccessError: <unprintable PathAccessError object>
could not access 'a', part 0 of Path('a', 'b'), got error: KeyError('a')
```

just need to figure out that `<unprintable XXX object>` bit

**Comment by kurtbrose** [comment]
alright this is getting pretty close now -- needs a closer look possibly at the format, and also may need to get a little creative to get the lines under test

**Comment by kurtbrose** [comment]
also, using `from None` exception raising which is python 3 only

**Comment by kurtbrose** [comment]
this is basically ready to go

File: glom/core.py [hunk 1]
```diff
@@ -100,7 +107,42 @@ class GlomError(Exception):
     (e.g., ``len``, ``sum``, any ``lambda``) will not be wrapped in a
     GlomError.
     """
-    pass
+    @classmethod
+    def wrap(cls, exc):
+        class GlomWrapError(type(exc), GlomError): pass
```
Comment (mahmoud) [inline glom/core.py:112]: Ah, that classic pattern going all the way back to the ASF days :)

File: glom/core.py [hunk 2]
```diff
@@ -100,7 +107,42 @@ class GlomError(Exception):
...
+        return GlomWrapError(*exc.args)
+        # this is another pattern, but doesn't work for C-defined exceptions
+        # (maybe if hasattr(sys.modules[type(exc).__module__], "__file__")?)
+        exc2 = copy.copy(exc)
+        exc2.__class__ = GlomWrapError
+        return exc2
```
Comment (mahmoud) [inline glom/core.py:118]: Is this cleaner in any way? Probably better to have just one way, right?

File: glom/core.py [hunk 3]
```diff
@@ -100,7 +107,42 @@ class GlomError(Exception):
...
+        from . import trace
+
+        # tried MANY approaches here -- one problem pragmatically is
+        # that traceback module and sys.exc_info() see different
+        # stacks here -- maybe it is something to do with running
+        # under pytest
+        lines = traceback.format_exc().split("\n")
+        limit = 0
+        for line in reversed(lines):
+            if __file__ in line and "_glom" in line:
```
Comment (mahmoud) [inline glom/core.py:130]: When is `__file__` in line but `_glom` is not?

Comment (mahmoud) [reply glom/core.py:130]: Also does this handle going in and out of glom? (e.g., callables which call glom themselves? also customized type handlers via register_op)

File: glom/core.py [hunk 4]
```diff
@@ -100,7 +107,42 @@ class GlomError(Exception):
...
+            limit += 1
+
+        self.short_stack = trace.short_stack(scope)
+        self.inner_format = ("\n".join(
+            ["", trace.short_stack(scope)] + lines[-limit:]))
+
+    def __str__(self):
+        if hasattr(self, "inner_format"):
+            return self.inner_format
+        return ""
```
Comment (mahmoud) [inline glom/core.py:142]: empty until finalized?

File: glom/core.py [hunk 5]
```diff
@@ -100,7 +107,42 @@ class GlomError(Exception):
...
+
+    def __repr__(self):
+        return str(self)
```
Comment (mahmoud) [inline glom/core.py:145]: Why have `__repr__` be the same as `__str__`? Exceptions often have these differ (notably, `KeyError`, for better or worse).

File: glom/core.py [hunk 6]
```diff
@@ -145,8 +187,9 @@ def __repr__(self):
         return '%s(%r, %r, %r)' % (cn, self.exc, self.path, self.part_idx)
 
     def __str__(self):
-        return ('could not access %r, part %r of %r, got error: %r'
-                % (self.path.values()[self.part_idx], self.part_idx, self.path, self.exc))
+        return GlomError.__str__(self) + (
+            'could not access %r, part %r of %r, got error: %r'
+            % (self.path.values()[self.part_idx], self.part_idx, self.path, self.exc))
```
Comment (mahmoud) [inline glom/core.py:192]: I guess super() might be a bit vague with this much multiple inheritance, is that right?

File: glom/core.py [hunk 7]
```diff
@@ -1199,19 +1242,19 @@ def _t_eval(target, _t, scope):
             try:
                 cur = getattr(cur, arg)
             except AttributeError as e:
-                raise PathAccessError(e, Path(_t), i // 2)
+                raise PathAccessError(e, Path(_t), i // 2) from None
```
Comment (mahmoud) [inline glom/core.py:1245]: I don't think this syntax is py2 compatible, but I take it you want to clear out the chained exception?

File: glom/core.py [hunk 8]
```diff
@@ -1821,14 +1864,24 @@ def glom(target, spec, **kwargs):
...
+                raise
+            ret = default
+    except Exception as e:
+        if isinstance(e, GlomError):
+            err = e
+        else:
+            err = GlomError.wrap(e)
+        err._finalize(scope[ERROR_SCOPE])
+    if err:
+        raise err
```
Comment (mahmoud) [inline glom/core.py:1884]: and we've got to play this game because python 3 has some weird scoping issue, right?

File: glom/core.py [hunk 9]
```diff
@@ -1840,12 +1893,16 @@ def _glom(target, spec, scope):
...
-    return scope[MODE](target, spec, scope)
+    try:
+        if isinstance(spec, TType):  # must go first, due to callability
+            return _t_eval(target, spec, scope)
+        elif callable(getattr(spec, 'glomit', None)):
+            return spec.glomit(target, scope)
+
+        return scope[MODE](target, spec, scope)
+    except Exception:
+        scope[ROOT].setdefault(ERROR_SCOPE, scope)
```
Comment (mahmoud) [inline glom/core.py:1904]: setdefault without assignment eh? I guess it's fewer characters than the alternative.

File: glom/test/test_trace.py [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+from glom.trace import line_stack, short_stack, tall_stack
+
+
+from glom.tutorial import CONTACTS
+
+
+def test_line_trace():
+    stacklifier = ([{'data': S}],)
+    scope = glom([1], stacklifier)[0]['data']
+    line_stack(scope)
```
Comment (mahmoud) [inline glom/test/test_trace.py:15]: are we going to want to test these against some values?

File: glom/trace.py [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+    # of execution
+    return [(scope, scope[Spec], scope[T]) for scope in list(reversed(scope.maps[:-2]))]
+
+
+# TODO: get fancier here and replace repr()
+# with something that only recurses one level
+def _format_value(value, maxlen):
+    s = repr(value)
+    if len(s) > maxlen:
+        s = s[:maxlen] + '...'
```
Comment (mahmoud) [inline glom/trace.py:34]: Agree we should do something fancier, but for now I'd do `max(maxlen - 3, 0)` at least

File: glom/trace.py [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+    return s
+
+
+def line_stack(scope):
+    """
+    unpack a scope into a single line summary
+    (shortest summary possible)
+    """
+    # the goal here is to do a kind of delta-compression --
```
Comment (mahmoud) [inline glom/trace.py:43]: delta compression is pretty fancy talk for "DRY out duplicate targets" ;)

File: glom/trace.py [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+    # if the target is the same, don't repeat it
+    segments = []
+    prev_target = _NO_TARGET
+    for scope, spec, target in _unpack_stack(scope):
+        segments.append('/')
```
Comment (mahmoud) [inline glom/trace.py:48]: interesting, slashes as the meta separator.

File: glom/core.py (hunk 2 above)
Comment (kurtbrose) [reply glom/core.py:118]: nope, my exception-fu is a little weak here

Comment (kurtbrose) [reply glom/core.py:118]: I'm thrashing around a bit

File: glom/core.py (hunk 3 above)
Comment (kurtbrose) [reply glom/core.py:130]: so, the goal here is "everything between the glom.glom() call at the top, and the last glom._glom() call at the bottom is replaced with glom stack"

in general I was not trying to handle cases where there is a nested call out to glom() again

I think that would probably be fine though, since the top-level glom() will intercept and re-raise -- from that point on glom._glom() is no longer in the traceback so it won't clip the stack trace too early

not sure how customized type handlers would enter the picture?

I think a dotted-line between glom and not-glom would work fine from a stack-trace coalescing perspective

File: glom/core.py (hunk 5 above)
Comment (kurtbrose) [reply glom/core.py:145]: I have no idea -- I was just trying to get the stack trace and repr to look right so I could do testing :-)

File: glom/test/test_trace.py (hunk 10 above)
Comment (kurtbrose) [reply glom/test/test_trace.py:15]: I'm not sure what the invariant is other than just grab the current value

File: glom/trace.py (hunk 11 above)
Comment (kurtbrose) [reply glom/trace.py:34]: oh - 3 is a good idea

File: glom/trace.py (hunk 12 above)
Comment (kurtbrose) [reply glom/trace.py:43]: I think of delta compression as "only show the value when it has changed"

Comment (kurtbrose) [reply glom/trace.py:43]: but its been a long time messing with these things, I might be out of it on wording

Comment (mahmoud) [reply glom/trace.py:43]: Yeah, I was just goofin anyways

File: glom/trace.py (hunk 13 above)
Comment (kurtbrose) [reply glom/trace.py:48]: yea, `short_stack` is the main one -- since that is what is used when glom() formats a stack trace

File: glom/test/test_error.py [hunk 14]
```diff
@@ -0,0 +1,18 @@
+
+import pytest
+
+from glom import glom, T, GlomError
+
+@pytest.mark.xfail
```
Comment (mahmoud) [inline glom/test/test_error.py:6]: I think these might actually be more of a place for `with pytest.raises` than xfails.

File: glom/core.py (hunk 7 above)
Comment (mahmoud) [reply glom/core.py]: fixed this, I think

File: glom/test/test_trace.py (hunk 10 above)
Comment (mahmoud) [reply glom/test/test_trace.py]: I added in what I meant, just that line_stack(scope) == "the expected string"

File: glom/core.py (hunk 8 above)
Comment (kurtbrose) [reply glom/core.py:1929]: without this, the whole stack trace is displayed even though it is a different error somehow

**Comment by mahmoud** [comment]
* Better line width default estimate
* Maybe play with reprlib
* Clean up trace.py and merge test_error and test_trace

**Comment by mahmoud** [comment]
(deferring on reprlib and numbering)

File: glom/core.py [hunk 15]
```diff
@@ -100,10 +115,126 @@ class GlomError(Exception):
...
+        # TODO: need to test this against a wide array of exception types
+        # this approach to wrapping errors works for exceptions
+        # defined in pure-python as well as C
+        exc_type = type(exc)
+        bases = (GlomError,) if issubclass(GlomError, exc_type) else (exc_type, GlomError)
+        exc_wrapper_type = type("GlomError.wrap({})".format(exc_type.__name__), bases, {})
+        try:
+            return exc_wrapper_type(*exc.args)
+        except Exception:  # maybe exception can't be re-created
+            return exc
```
Comment (mahmoud) [inline glom/core.py:129]: this approach looks like it'll work. i hope it's not adding much overhead, but it's the exception path. then again, exception path gets hit pretty often for folks using defaults and stuff. 

eh, we'll benchmark it when we benchmark it. i think this is probably even a little safer than we need to be. if we end up optimizing this away later, once we know how corner casey this is, so be it.

**Comment by mahmoud** [comment]
OK I was almost right with my approval message. Hit a couple missed lines and added a unicode test for good measure. Merging!
